src/osd/ReplicatedPG.cc

```diff
@@ -1496,6 +1496,22 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // discard due to cluster full transition?  (we discard any op that
+  // originates before the cluster or pool is marked full; the client
+  // will resend after the full flag is removed or if they expect the
+  // op to succeed despite being full).
+  // FIXME: we exclude mds writes for now.
+  if (!m->get_source().is_mds() &&
+      info.history.last_epoch_marked_full > m->get_map_epoch()) {
+    dout(10) << __func__ << " discarding op sent before full " << m << " "
+	     << *m << dendl;
+    return;
+  }
+  if (osd->check_failsafe_full()) {
+    osd->reply_op_error(op, -ENOSPC);
```

Comment:
why return the error here (and in prepare_transaction()) rather than dropping the request? doesn't check_failsafe_full() imply we'll mark the cluster full soon, or go down, making the client resend either way?